### ISSUE TITLE
[IDEA] Draggable rows macro ( - table rows?)

### ISSUE THREAD

**Issue by twMat** [TiddlyWiki/TiddlyWiki5#5554]
**Is your feature request related to a problem? Please describe.**
[Ref discussion](https://groups.google.com/g/tiddlywiki/c/WAWgE3hmBiM)
I want to easily create **DnD-rearrangable rows**, i.e where each row has **complex data**, not just a single title.
The listed output needs to be presented neatly. For example:

HelloThere   &nbsp;  TableOfContents &nbsp;&nbsp;&nbsp;  DefaultTiddler
Learning   &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;     TableOfContents  &nbsp;&nbsp;&nbsp;  NotDefault

**Describe the solution you'd like**
For a neat result, the obvious solution is a *table*. So, each *table row* should be rearrangable. I hope for something comparable to the [list-links-draggable](https://tiddlywiki.com/prerelease/#list-links-draggable%20Macro) macro.

**Describe alternatives you've considered**
I have tried to use the list-links-draggable macro like so: 

```
text:
<table>
<<list-links-draggable type:table subtype:tr itemTemplate:mytemplate>>
</table>
list: 
one two three four
```

```
title: mytemplate
text: <td><$link/></td>
```

...and it does output a table but it is ruined by extra divs, i.e this structure:

```
<table>
<tr class="tc-droppable">
   <div class="tc-droppable-placeholder"></div>
   <div><td>...first item..</td></div>
</tr>
<tr class="tc-droppable">
   <div class="tc-droppable-placeholder"></div>
   <div><td>...second item..</td></div>
</tr>
...etc...
</table>
```

**Additional context**
One more thing... if the output indeed is in a table it'd be nice if the table has a default style class like `class="draggable-table"` or some such.

**Comment by pmario** [comment]
IMO you should have a look at the list-links-draggable code and change it to list-table-dragable. So you can decide it there should be extra divs or TDs TRs ... or what ever you need

**Comment by saqimtiaz** [comment]
Also worth noting that wikitext macros are often a convenience shortcut to using widgets for oft repeated patterns. So working directly with the draggable and droppable widgets, you have complete flexibility in how to implement this.

To me this is a good example of something that can be explored via a wikitext plugin, and if a solution is found that is generic enough, it could then potentially be considered for the core. Having worked with draggable rows in tables a few times, I've found that a one size fits all solution is rarely good enough, and depending on the desired table layout and workflow desired, customization is often required.

**Comment by twMat** [comment]
I just created 

https://list-table-draggable.tiddlyhost.com/

I hope something similar can be included in standard distro or as official plugin. It makes sense to enable the list items to be more complex than merely showing the title as with the other draggable macros.

**Comment by pmario** [comment]
Cool stuff! But I think, there should be handles at the beginning or the end of the row. like so:

![grafik](https://user-images.githubusercontent.com/374655/115966260-fdcac300-a52c-11eb-95e4-678b7e950c33.png)

**Comment by saqimtiaz** [comment]
Yep, a solution for the core needs to be more flexible so you don't have to rely on having links in each row to drag. 

Overall, this just highlights the need to refactor `list-links-draggable` to make it more flexible to accommodate table layouts, and potentially add a template that can be used for table layout.

**Comment by twMat** [comment]
Yes, it should really be the *whole* item that is listed that should be draggable - in the case of my macro it'd be the `tr` element.

@pmario , it is simple to create a handle in the template like so `<td><$link>::</$link></td>`. I guess you could style it if you don't want it blue. I should add this to the demos, thanks for bringing it up.


@saqimtiaz 
>Overall, this just highlights the need to refactor `list-links-draggable` to make it more flexible to accommodate table layouts, and potentially add a template that can be used for table layout.

+1 ...But where do I go with a comment like that? Should I post an issue request?

**Comment by kookma** [comment]
> Overall, this just highlights the need to refactor `list-links-draggable` to make it more flexible to accommodate table layouts, and potentially add a template that can be used for table layout.

That's quite true!
For example it can accept type=table, subtype=tr and generate a draggable table

```
<<list-links-draggable tiddler:"Days of the Week" type:table subtype:tr itemTemplate:myTemplate>>
```

Also a handle would be nice as Mario mentioned!

**Comment by twMat** [comment]
Did I close this? (Why?)

I wish to easily create DnD-rearrangable rows, i.e where each row has complex data, not just a single title.

bump

**Comment by twMat** [comment]
For the record, I've now made the [Tids-to-Table](https://tids-to-table.tiddlyhost.com/) plugin (replacing earlier variants).

Considering the popularity of *tables*, I still think the core should enable something like this. Actually, if *list-links-draggable* didn't exist and someone proposed it, I would expect the reply to be *"Rearranging only bulleted links is too limited - make it more general!"*. BTW, as someone noted, because my plugin produces the rows from a *filter* it has similarities to SQL tables.

**Comment by twMat** [comment]
@pmario thank you - I'm replying [here](https://talk.tiddlywiki.org/t/presenting-tids-to-table-generate-tables-from-filters/7517/11?u=twmat).
